### ISSUE TITLE
Error when dealing with nested subset types

### ISSUE THREAD

**Issue by tardism** [dafny-lang/dafny#2639]
The following is a simple reporoduction

```dafny
type Foo = s: string | |s|!=0 witness *
type Bar = y: Foo | |y| == 2 

// The error is here
function Baz():bool{
    true
}
```

```
[{
	"resource": "/c:/Users/Tardi/OneDrive/文档/Dafny-Programs/SubsetOfSubsetBug.dfy",
	"owner": "dafny-vscode",
	"severity": 8,
	"message": "Dafny encountered an internal error. Please report it at <https://github.com/dafny-lang/dafny/issues>.
System.AggregateException: One or more errors occurred. (Unable to cast object of type 'Microsoft.Dafny.TypeSynonymDecl' to type 'Microsoft.Dafny.NewtypeDecl'.)
 ---> System.InvalidCastException: Unable to cast object of type 'Microsoft.Dafny.TypeSynonymDecl' to type 'Microsoft.Dafny.NewtypeDecl'.
   at Microsoft.Dafny.Translator.CheckResultToBeInType_Aux(IToken tok, Expression expr, Type toType, BoogieStmtListBuilder builder, ExpressionTranslator etran, String errorMsgPrefix)
   at Microsoft.Dafny.Translator.CheckResultToBeInType_Aux(IToken tok, Expression expr, Type toType, BoogieStmtListBuilder builder, ExpressionTranslator etran, String errorMsgPrefix)
   at Microsoft.Dafny.Translator.CheckResultToBeInType(IToken tok, Expression expr, Type toType, List`1 locals, BoogieStmtListBuilder builder, ExpressionTranslator etran, String errorMsgPrefix)
   at Microsoft.Dafny.Translator.AddWellformednessCheck(RedirectingTypeDecl decl)
   at Microsoft.Dafny.Translator.AddTypeDecl(SubsetTypeDecl dd)
   at Microsoft.Dafny.Translator.DoTranslation(Program p, ModuleDefinition forModule)
   at Microsoft.Dafny.Translator.Translate(Program p, ErrorReporter reporter, TranslatorFlags flags)+MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location ---
   at Microsoft.Dafny.LanguageServer.Language.DafnyProgramVerifier.GetVerificationTasksAsync(DafnyDocument document, CancellationToken cancellationToken)
   at Microsoft.Dafny.LanguageServer.Workspace.TextDocumentLoader.PrepareVerificationTasksAsync(DafnyDocument loaded, CancellationToken cancellationToken)
   at Microsoft.Dafny.LanguageServer.Workspace.DocumentDatabase.LoadVerificationTasksAsync(Task`1 resolvedDocumentTask, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---",
	"source": "Crash",
	"startLineNumber": 4,
	"startColumn": 10,
	"endLineNumber": 4,
	"endColumn": 13
}]

```

**Comment by cpitclaudel** [comment]
Possible duplicate of https://github.com/dafny-lang/dafny/issues/2253 ?

**Comment by tardism** [comment]
This may be true.
However the above example will work if you add `witness *` to `Bar`
Either way, related

**Comment by MikaelMayer** [comment]
Fixed by #2954
